projects/debouncer/src/debouncer.c

```diff
@@ -0,0 +1,54 @@
+#include "debouncer.h"
+
+#include <stddef.h>
+#include "soft_timer.h"
+
+#define DEBOUNCER_INTERRUPT_MASKING_DURATION_MS 50
+
+// This is the callback for the soft timer. If there is a button input, it runs the user's callback.
+static void prv_timer_callback(SoftTimerID timer_id, void *context) {
+  DebouncerInfo *debouncer_info = context;
+
+  GPIOState current_state;
+  gpio_get_state(&debouncer_info->address, &current_state);
+  if (!debouncer_info->callback) return;
+  if (current_state == debouncer_info->state) {
+    debouncer_info->callback(&debouncer_info->address, debouncer_info->context);
+  }
+
+  gpio_it_mask_interrupt(&debouncer_info->address, false);
+}
+
+// This is the interrupt callback to start off the debouncing
+static void prv_it_callback(const GPIOAddress *address, void *context) {
+  DebouncerInfo *debouncer_info = context;
+  gpio_get_state(address, &debouncer_info->state);
+
+  gpio_it_mask_interrupt(address, true);
+
+  soft_timer_start_millis(DEBOUNCER_INTERRUPT_MASKING_DURATION_MS, prv_timer_callback,
+                          debouncer_info, NULL);
+}
+
+StatusCode debouncer_init_pin(DebouncerInfo *debouncer_info, const GPIOAddress *address,
+                              gpio_it_callback callback, void *context) {
+  GPIOSettings gpio_settings = {
+    .direction = GPIO_DIR_IN,   //
```

Comment:
With clang-format v5 trailing `//` isn't needed anymore.

Also doing this will override the existing settings on the pin. Personally I think configuring the pin shouldn't be implicit to the debouncer but I'll defer to @tituschow.

Reply:
Hmm, I've still needed trailing `//` to keep it from putting it on one line.

I think it makes sense to override it to an input because there's no other valid setting.

Reply:
So we're good on this? :))